File: packages/guards/src/number-precision.ts
import type { Guard, GuardResult, GuardContext } from 'open-guardrail-core';

interface NumberPrecisionOptions {
  action: 'block' | 'warn';
  maxDecimalPlaces?: number;
}

const DECIMAL_RE = /\b\d+\.(\d+)\b/g;

export function numberPrecision(options: NumberPrecisionOptions): Guard {
  const maxDp = options.maxDecimalPlaces ?? 6;

  return {
    name: 'number-precision',
    version: '0.1.0',
    description: 'Validates number precision in output',
    category: 'format',
    supportedStages: ['output'],
    async check(text: string, _ctx: GuardContext): Promise<GuardResult> {
      const start = performance.now();
      const issues: string[] = [];
      const decimalLengths: number[] = [];

      DECIMAL_RE.lastIndex = 0;
      let match: RegExpExecArray | null;
      while ((match = DECIMAL_RE.exec(text)) !== null) {
        const dp = match[1].length;
        decimalLengths.push(dp);
        if (dp > maxDp) {
          issues.push(`excessive_precision:${match[0]}(${dp}dp)`);
        }
      }

      if (decimalLengths.length >= 2) {
        const unique = new Set(decimalLengths);
        const range = Math.max(...decimalLengths) - Math.min(...decimalLengths);
        if (unique.size > 2 && range > 3) {
          issues.push('inconsistent_precision');
        }
      }

      const triggered = issues.length > 0;
      const score = triggered ? Math.min(issues.length / 3, 1.0) : 0;

      return {
        guardName: 'number-precision',
        passed: !triggered,
        action: triggered ? options.action : 'allow',
        score,
        latencyMs: Math.round(performance.now() - start),
        details: triggered ? { issues, maxDecimalPlaces: maxDp } : undefined,
      };
    },
  };
}

File: python/open_guardrail/guards/number_precision.py
"""Validate number precision in output."""
from __future__ import annotations

import math
import re
import time

from open_guardrail.core import GuardResult

_DECIMAL_RE = re.compile(r"\b\d+\.(\d+)\b")


class _NumberPrecision:
    def __init__(
        self,
        *,
        action: str = "warn",
        max_decimal_places: int = 6,
    ) -> None:
        self.name = "number-precision"
        self.action = action
        self._max_dp = max_decimal_places

    def check(
        self, text: str, stage: str = "output"
    ) -> GuardResult:
        start = time.perf_counter()
        issues: list[str] = []
        dp_list: list[int] = []

        for match in _DECIMAL_RE.finditer(text):
            dp = len(match.group(1))
            dp_list.append(dp)
            if dp > self._max_dp:
                issues.append(
                    f"excessive_precision:"
                    f"{match.group(0)}({dp}dp)"
                )

        if len(dp_list) >= 2:
            unique = set(dp_list)
            rng = max(dp_list) - min(dp_list)
            if len(unique) > 2 and rng > 3:
                issues.append("inconsistent_precision")

        triggered = len(issues) > 0
        score = (
            min(len(issues) / 3, 1.0)
            if triggered
            else 0.0
        )
        elapsed = (time.perf_counter() - start) * 1000

        return GuardResult(
            guard_name="number-precision",
            passed=not triggered,
            action=self.action if triggered else "allow",
            score=score,
            message=(
                "Number precision issue detected"
                if triggered
                else None
            ),
            latency_ms=round(elapsed, 2),
            details=(
                {
                    "issues": issues,
                    "max_decimal_places": self._max_dp,
                }
                if triggered
                else None
            ),
        )


def number_precision(
    *,
    action: str = "warn",
    max_decimal_places: int = 6,
) -> _NumberPrecision:
    return _NumberPrecision(
        action=action,
        max_decimal_places=max_decimal_places,
    )
